cardano-api/src/Cardano/Api/Query.hs

```diff
@@ -88,6 +89,14 @@ data QueryInMode mode result where
 
 deriving instance Show (QueryInMode mode result)
 
+createQueryInMode :: EraInMode era mode
+                  -> QueryInShelleyBasedEra era result
+                  -> QueryInMode mode (Either EraMismatch result)
+createQueryInMode eraInMode' query =
+  case cardanoEraStyle $ eraInModeToEra eraInMode' of
+    LegacyByronEra -> error $ "Expecting shelley based era when trying to create query. \
```

Comment:
@dcoutts it would be nice if I could figure out how to tell the compiler the `LegacyByronEra` case isn't possible. I tried `case eraInMode of {}` but that didn't work :thinking: